com.ibm.streamsx.inet/com.ibm.streamsx.inet.http/HTTPBulkRequest.spl

```diff
@@ -0,0 +1,125 @@
+ * @param	width				The number of parallel used threads. This conforms to the maximum number if used http connections. Default is 8
+ * @param	queue				The length of the input queue ot each thread. Default 1
+ * @param	congestionPolicy	The congestion policy of the input queue. Type *Sys.CongestionPolicy*. Default Sys.Wait
+ * @param	user				If this string is not empty, this user name is passed to to the server.
+ * @param	password			If this string is not empty, this password is passed to the server
+ * @param	certFile			If this string is not empty, this is the absolute path to a certificate file sent to the server
+ * @param	certType			If this string is not empty, this is the cert file type (PEM, P12)
+ * @param	keyFile				If this string is not empty, this is the absolute path to a private key file to be sent to the server
+ * @param	keyType				If this string is not empty, this is the key file type (PEM, P12)
+ * @param	keyPass				If this string is not empty, this is the password for the private key
+ */
+public composite HTTPBulkRequest (output Result, Error; input Request) {
+	param
+		expression<HttpMethod>    $method;
+		attribute                 $url;
+		function                  $urlEncode : urlEncode;
+		expression<rstring>       $data : "";
+		expression<list<rstring>> $headers : [];
+		expression<rstring>       $contentTypeHeader: "ContentType: application/x-www-form-urlencoded";
+		expression<int32>         $requestTimeout : 5;
+		expression<int32>         $connectionTimeout : 120;
+		expression<int32>         $width: 8;
+		expression<int32>         $queue : 1;
+		expression<Sys.CongestionPolicy> $congestionPolicy : Sys.Wait;
+		expression<rstring>       $user : "";
+		expression<rstring>       $password : "";
+		expression<rstring>       $certFile : "";
+		expression<rstring>       $certType : "";
+		expression<rstring>       $keyFile : "";
+		expression<rstring>       $keyType : "";
+		expression<rstring>       $keyPass : "";
+
+	type
+		/**
+		 * This type determines the kind of the http operation
+		 * values are httpGet, httpPut, httpPost, httpDelete
+		 */
+		static HttpMethod = enum {httpGet, httpPut, httpPost, httpDelete};
+
+	graph
+		@parallel(width=$width)
+		(
+			stream<I, tuple<rstring body, list<rstring> headers, rstring message, int32 status, rstring url>> Result as R;
+			stream<rstring url, rstring message, int32 status> Error as E
+		) = Custom(Request as I) {
+			logic
+				onTuple I: {
+					mutable int32 error_ =0;
+					mutable list<rstring> replyHeaders_ = [];
+					mutable list<rstring> requestHeaders_ = $headers;
+					if (length($contentTypeHeader) > 0) {
+						appendM(requestHeaders_, $contentTypeHeader);
+					}
+					rstring urlProcessed_ = $urlEncode($url);
+					mutable rstring result_ = "";
+					appTrc(Trace.trace, "Emit http request "+
```

Comment:
Should these traces be protected with an if statement to stop the string being created when the tracing is not enabled, or does that not matter for SPL?

Reply:
good point. I will change this